File: apps/mobile/lib/running.ts
import { apiRequest } from './apiRequest';
import type { TokenGetter } from './useAuthToken';

/**
 * The running half of a session — client half of `internal/modules/running`'s
 * session files.
 *
 * A running session is an ordinary session (`sport: 'running'`) plus this,
 * mirroring `lib/bjjSession.ts`'s relationship to `internal/modules/bjj`
 * exactly, for the same reason: keeping the session row where every other
 * sport's lives is what makes a run show up in training history, the
 * consistency grid and the cross-sport load picture, instead of in a corner
 * of the app labelled Running. This module owns only what a run has and a
 * lift or a mat session do not: the GPS track, splits, elevation and pace.
 */

/** Where the track and numbers came from. Matches `running.Source` exactly. */
export type Source = 'phone_gps' | 'healthkit' | 'manual';

/**
 * One recorded point along the run.
 *
 * `elevation_m` is nullable — common indoors, on an older phone, or on a
 * thinned-out imported track — and `null` is deliberately not defaulted to 0,
 * which would assert sea level for a run that simply didn't say.
 */
export type RoutePoint = {
  lat: number;
  lng: number;
  elevation_m: number | null;
  /** ISO 8601 / RFC3339 — what lets a client re-derive pace-over-time. */
  recorded_at: string;
};

/** One distance-based split — "this kilometre took 5:12". */
export type Split = {
  distance_m: number;
  duration_seconds: number;
};

export type SessionDetail = {
  session_id: string;
  route_points: RoutePoint[];
  splits: Split[];
  elevation_gain_m: number | null;
  avg_pace_sec_per_km: number | null;
  distance_m: number | null;
  duration_seconds: number | null;
  source: Source;
  created_at?: string;
  updated_at?: string;
};

/** A blank detail for a freshly-started run, tracked live by this phone. */
export function emptyDetail(sessionID: string): SessionDetail {
  return {
    session_id: sessionID,
    route_points: [],
    splits: [],
    elevation_gain_m: null,
    avg_pace_sec_per_km: null,
    distance_m: null,
    duration_seconds: null,
    source: 'phone_gps',
  };
}

/**
 * The exercise id the generic personal-record pipeline reads for a run.
 *
 * Matches `backend/internal/modules/exercise/exercises.json`'s `"run"` entry
 * (`load_type: distance_time`, `sport: running`) — a running session writes a
 * `session_sets` row against this id so `longest_time`/`furthest_distance`
 * already work for it through `session.Records`, exactly as
 * `internal/modules/running/running.go`'s package doc describes. Duplicated
 * as a constant rather than derived, the same call `running.go` itself makes
 * for `sportKey`: this is a storage-level fact about the seeded catalog, and
 * a client-side registry lookup would run the dependency backwards.
 */
export const RUN_EXERCISE_ID = 'run';

/* ---------------------------------------------------------------------------
 * Pure calculation from a GPS track.
 *
 * Everything below takes only plain data (points, splits, numbers) and
 * returns plain data — no SQLite, no network, no React. That is what makes it
 * testable with a fixture track and nothing else running.
 * ------------------------------------------------------------------------- */

const EARTH_RADIUS_M = 6371000;

/**
 * Great-circle distance between two points, in metres.
 *
 * The haversine formula — accurate enough for a run-length track (it starts
 * to matter at hundreds of kilometres, not kilometres) and cheap enough to
 * call once per consecutive point pair on a phone.
 */
export function haversineMeters(
  a: { lat: number; lng: number },
  b: { lat: number; lng: number },
): number {
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const h = Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  // Clamped: floating-point drift can push `h` fractionally past 1, which
  // makes `Math.sqrt` produce a value `asin` cannot accept and returns NaN
  // for two points that are, for any practical purpose, identical.
  return 2 * EARTH_RADIUS_M * Math.asin(Math.min(1, Math.sqrt(h)));
}

/**
 * Total distance covered by a track, in metres.
 *
 * The sum of consecutive great-circle segments — not a straight line from
 * start to finish, which would silently shrink every out-and-back and every
 * loop to a fraction of what was actually run.
 */
export function trackDistanceMeters(points: RoutePoint[]): number {
  let total = 0;
  for (let i = 1; i < points.length; i++) total += haversineMeters(points[i - 1], points[i]);
  return total;
}

/**
 * Wall-clock span of a track, from its first recorded point to its last, in
 * seconds.
 *
 * **This is span, not ACTIVE time, and the two differ the moment a run is
 * paused.** A pause stops the live screen recording new points for the
 * paused interval, but a track's first and last points still bound the
 * WHOLE run including that gap — this function has no way to know a gap in
 * the middle was a pause rather than a dead GPS patch, and must not guess.
 * So the live screen does NOT use this to compute the `duration_seconds` it
 * sends to the server: it keeps its own active-time accumulator (ticking
 * only between Resume and the next Pause/Finish) for that, and reserves this
 * function for a track known to have no pauses in it — the common shape a
 * fixture test exercises, and a manual/imported track with no live pausing
 * at all.
 */
export function trackDurationSeconds(points: RoutePoint[]): number {
  if (points.length < 2) return 0;
  const first = new Date(points[0].recorded_at).getTime();
  const last = new Date(points[points.length - 1].recorded_at).getTime();
  return Math.max(0, (last - first) / 1000);
}

/**
 * Seconds per kilometre, averaged over a distance and a duration.
 *
 * `null` for a non-positive distance rather than `Infinity` or a divide-by-
 * zero `NaN` — "no pace yet" is a state every caller (the live readout, the
 * finished-run summary) has to render as a dash, and a numeric sentinel would
 * make each one re-derive that check independently.
 */
export function averagePaceSecPerKm(distanceM: number, durationSeconds: number): number | null {
  if (!(distanceM > 0) || !(durationSeconds >= 0)) return null;
  return durationSeconds / (distanceM / 1000);
}

/**
 * Total climb, in metres — the sum of positive elevation deltas between
 * consecutive points.
 *
 * Descent is not subtracted (this is GAIN, not net elevation change), and a
 * point missing `elevation_m` contributes no delta on either side of it
 * rather than being treated as a drop to/from zero — a single dropped
 * reading must not register as a cliff.
 */
export function elevationGainMeters(points: RoutePoint[]): number {
  let gain = 0;
  for (let i = 1; i < points.length; i++) {
    const a = points[i - 1].elevation_m;
    const b = points[i].elevation_m;
    if (a == null || b == null) continue;
    if (b > a) gain += b - a;
  }
  return gain;
}

/** The split distance every running app and every runner already thinks in. */
export const DEFAULT_SPLIT_METERS = 1000;

/**
 * Distance-based splits derived from a timestamped track.
 *
 * Walks the track's cumulative distance and, each time it crosses a split
 * boundary, LINEARLY INTERPOLATES the crossing time between the two points
 * that straddle it — both on distance and on time — rather than rounding to
 * the nearest recorded point. A GPS fix lands wherever the phone's sample
 * interval happens to put it, essentially never exactly on a kilometre; a
 * split reported as "whichever point was closest" would be off by however
 * long that interval is, every single split.
 *
 * A track with two or more boundaries inside one segment (a very sparse
 * track, or a `splitMeters` smaller than the sample spacing) still gets one
 * split per boundary crossed — the inner `while` keeps interpolating from the
 * same segment until it runs out of boundaries to cross.
 *
 * **Known, narrow, and out of scope**: if a split boundary happens to fall
 * inside the segment spanning a pause (no points are recorded while paused,
 * so that segment's two endpoints straddle the whole gap), that ONE split's
 * `duration_seconds` is inflated by however long the pause lasted — the same
 * limitation `trackDurationSeconds` documents for the run as a whole, here at
 * the scale of a single split rather than the total. Rare in practice (it
 * requires pausing within the last stretch of a kilometre) and not corrected
 * here, the same way `trackDurationSeconds` is not.
 */
export function splitsFromTrack(points: RoutePoint[], splitMeters = DEFAULT_SPLIT_METERS): Split[] {
  if (points.length < 2 || !(splitMeters > 0)) return [];

  const splits: Split[] = [];
  let cumulativeDistance = 0;
  let splitStartDistance = 0;
  let splitStartTime = new Date(points[0].recorded_at).getTime();
  let nextBoundary = splitMeters;

  for (let i = 1; i < points.length; i++) {
    const prev = points[i - 1];
    const cur = points[i];
    const segDist = haversineMeters(prev, cur);
    if (segDist <= 0) continue;

    const segStartDist = cumulativeDistance;
    const prevTime = new Date(prev.recorded_at).getTime();
    const curTime = new Date(cur.recorded_at).getTime();
    cumulativeDistance += segDist;

    while (cumulativeDistance >= nextBoundary) {
      const fraction = (nextBoundary - segStartDist) / segDist;
      const boundaryTime = prevTime + fraction * (curTime - prevTime);
      splits.push({
        distance_m: nextBoundary - splitStartDistance,
        duration_seconds: Math.max(0, Math.round((boundaryTime - splitStartTime) / 1000)),
      });
      splitStartDistance = nextBoundary;
      splitStartTime = boundaryTime;
      nextBoundary += splitMeters;
    }
  }
  return splits;
}

export function putDetail(
  getToken: TokenGetter,
  sessionID: string,
  detail: SessionDetail,
): Promise<{ detail: SessionDetail }> {
  return apiRequest<{ detail: SessionDetail }>(
    getToken,
    `/running/sessions/${encodeURIComponent(sessionID)}`,
    { method: 'PUT', body: JSON.stringify(detail) },
  );
}

export function getDetail(
  getToken: TokenGetter,
  sessionID: string,
): Promise<{ detail: SessionDetail }> {
  return apiRequest<{ detail: SessionDetail }>(
    getToken,
    `/running/sessions/${encodeURIComponent(sessionID)}`,
  );
}

File: backend/internal/modules/running/running.go
// Package running holds the running half of a session.
//
// A running session is a row in `sessions` like any other — `sport =
// "running"` — so it stays visible to training history, the consistency
// grid and the cross-sport load currency exactly like a strength or BJJ
// session does. This package owns only what a run has and a lift does not:
// the GPS track, distance splits, elevation and pace.
//
// The split mirrors `bjj`'s SessionDetail deliberately: the client creates
// the session through the session module (a plain `sessions` row, and
// usually a `session_sets` row against a "run" exercise so the generic
// personal-record pipeline sees it — see the package doc on
// `RecordKindsFor`), then PUTs this module's own detail alongside it.
// Nothing here writes the `sessions` table, and nothing here computes a
// personal record: `longest_time` and `furthest_distance` already work for
// any distance_time exercise through `session.Records`, and pace-normalized
// PRs (fastest 5k, fastest 10k) are explicitly out of scope — see the
// history entry this package landed with.
package running

import (
	"context"
	"errors"
	"slices"
	"time"
)

var (
	// ErrNotFound covers "no such session", "not yours" and "not a running
	// session" alike — deliberately the same error everywhere, so a caller
	// cannot use the response to probe for session ids. Same reasoning as
	// bjj.ErrNotFound.
	ErrNotFound = errors.New("running: not found")
	// ErrInvalidInput covers both a value the domain rejects (Validate) and a
	// constraint Postgres rejects on our behalf (translatePgError).
	ErrInvalidInput = errors.New("running: invalid input")
	// ErrAlreadyExists means this HealthKit UUID is already attached to a
	// DIFFERENT session's DETAIL row for this user — the per-user unique
	// index on healthkit_uuid firing. PutDetail is an upsert on session_id,
	// so this can never mean "the same session was saved twice"; it means
	// two different sessions are claiming the same HealthKit workout.
	//
	// This index refuses the DUPLICATE DETAIL ROW, no more — by the time it
	// fires, the caller's generic `sessions` row (created through the
	// session module, before this endpoint is ever reached) already exists
	// server-side. Refusing the detail alone would leave that a real,
	// permanent, detail-less duplicate in the athlete's training history.
	// What makes the end-to-end guarantee "no duplicate RUN" hold is the
	// mobile client's own handling of this exact error: on a 409 here for a
	// HealthKit-sourced session, it deletes the session it just created
	// (locally and via a DELETE call) rather than leaving it orphaned — see
	// `abandonDuplicateHealthKitImport` in apps/mobile/lib/sessionStore.ts.
	// This index is the backstop for the case the import flow's own local
	// ledger cannot catch on its own (a reinstalled app or a second device,
	// neither of which has a ledger to consult) — see HealthKitUUID's doc
	// comment.
	ErrAlreadyExists = errors.New("running: already exists")
)

// sportKey is the `sessions.sport` value this module owns.
//
// Matches the registry key in `internal/platform/discipline`. Duplicated as
// a constant rather than imported, for the same reason bjj does it: the
// registry describes what a client may enable, and this is a storage-level
// invariant about which rows this module may write to — the dependency
// would otherwise run backwards.
const sportKey = "running"

// Source is where a run's track and numbers came from.
type Source string

const (
	// SourcePhoneGPS is a track recorded live by the phone during the run.
	SourcePhoneGPS Source = "phone_gps"
	// SourceHealthKit is a run imported from Apple Health / HealthKit —
	// recorded by a watch or another app, not this one.
	SourceHealthKit Source = "healthkit"
	// SourceManual is hand-entered: distance and duration typed in after the
	// fact, with no track at all. Still a complete session — not every run
	// happens with a phone along.
	SourceManual Source = "manual"
)

var sources = []Source{SourcePhoneGPS, SourceHealthKit, SourceManual}

// Sources lists the source vocabulary, for a client rendering a picker
// without hardcoding the list a second time.
func Sources() []Source {
	out := make([]Source, len(sources))
	copy(out, sources)
	return out
}

func (s Source) Valid() bool {
	return slices.Contains(sources, s)
}

// RoutePoint is one recorded point along the run.
//
// Timestamped rather than merely ordered: RecordedAt is what lets a client
// re-derive pace-over-time or draw the track against a clock, and an array
// index alone cannot survive a partial re-send the way a timestamp can.
type RoutePoint struct {
	Lat float64 `json:"lat"`
	Lng float64 `json:"lng"`
	// ElevationM is nil when the device did not report one — common indoors,
	// on an older phone, or on an imported track with a thinned-out feed.
	// Not defaulting to 0 avoids asserting sea level for a run that simply
	// didn't say.
	ElevationM *float64  `json:"elevation_m"`
	RecordedAt time.Time `json:"recorded_at"`
}

// maxLatDegrees and maxLngDegrees bound a coordinate to the range a real GPS
// fix can report. Not a precision check, and not a reliable catch for a
// swapped lat/lng either — most real coordinates fit in both ranges, so a
// swap only trips this for the minority that don't (e.g. anything with a
// longitude past ±90). What it does reliably catch is a value outside what
// a coordinate can physically be — a stray sentinel, a unit mixup — before
// it becomes a track nobody can render sensibly on a map.
const (
	maxLatDegrees = 90.0
	maxLngDegrees = 180.0
)

func (p RoutePoint) valid() bool {
	if p.Lat < -maxLatDegrees || p.Lat > maxLatDegrees {
		return false
	}
	if p.Lng < -maxLngDegrees || p.Lng > maxLngDegrees {
		return false
	}
	return true
}

// Split is one distance-based split — "this kilometre took 5:12".
//
// Distance-based rather than time-based because that is how every running
// app and every runner already thinks about a split; a time-based split
// ("every 5 minutes") is a different, rarer view this ticket does not build.
type Split struct {
	DistanceM       float64 `json:"distance_m"`
	DurationSeconds int     `json:"duration_seconds"`
}

func (s Split) valid() bool {
	return s.DistanceM > 0 && s.DurationSeconds > 0
}

// MaxRoutePoints bounds the track on one session.
//
// A GPS track sampled every few seconds on a two-hour long run is on the
// order of a thousand points; 20,000 is far past anything a real run
// produces and exists only to stop an unbounded client payload from becoming
// an unbounded transaction — the same reasoning as bjj.MaxTags.
const MaxRoutePoints = 20000

// MaxSplits bounds the split list. A marathon at 1km splits is 42; 500 is far
// past any real run, including one split per tenth of a kilometre on an
// ultra.
const MaxSplits = 500

// maxHealthKitUUIDLength bounds a HealthKit workout UUID. Not a format check
// (Validate does not parse this as a real UUID, the same "range, not shape"
// stance the rest of this file takes on a vocabulary it does not otherwise
// need to police) — HKWorkout.uuid is Apple's standard 36-character UUID
// string, and this is generous headroom over that rather than an exact fit,
// so it stops an absurd payload without becoming a second place the UUID
// format has to be kept in sync with reality.
const maxHealthKitUUIDLength = 128

// SessionDetail is everything a run has beyond the generic session row.
type SessionDetail struct {
	SessionID string `json:"session_id"`

	// RoutePoints is the GPS track, in recording order. Empty for a manual
	// entry or an imported summary with no track — a real run, not an error.
	RoutePoints []RoutePoint `json:"route_points"`
	// Splits are the distance-based splits, in order.
	Splits []Split `json:"splits"`

	// ElevationGainM is total climb, derived client-side from the track (or
	// reported by HealthKit) rather than recomputed here — this module
	// stores what it is told, the same stance session.Set.LoadFactor's
	// sibling fields take on derived numbers elsewhere in this codebase.
	ElevationGainM *float64 `json:"elevation_gain_m"`
	// AvgPaceSecPerKm is seconds per kilometre, averaged over the whole run.
	// Nil rather than computed from DistanceM/DurationSeconds here: a client
	// with a real track may have a better number (excluding paused time),
	// and this module must not silently override it with a coarser one.
	AvgPaceSecPerKm *float64 `json:"avg_pace_sec_per_km"`
	// DistanceM and DurationSeconds are the run's own numbers.
	//
	// Deliberately duplicated from wherever a `session_sets` row for this
	// session's exercise also carries them (see the package doc): that row
	// is what the generic Records pipeline reads for longest_time and
	// furthest_distance, while these are what THIS module's own detail
	// screen renders without a second fetch. The two are populated from the
	// same client-side numbers and are not reconciled against each other —
	// same relationship bjj's SessionRPE has to session_sets.rpe on a
	// strength session: same fact, two homes, because they serve different
	// screens.
	DistanceM       *float64 `json:"distance_m"`
	DurationSeconds *int     `json:"duration_seconds"`

	Source Source `json:"source"`

	// HealthKitUUID is the `HKWorkout.uuid` a HealthKit-imported run came
	// from — nil for every other source. This is the dedup key: N465's
	// import flow checks its own local ledger before ever reaching here, but
	// a reinstalled or second device has no local ledger to check, so this
	// column carries a per-user unique index (see the migration) as the
	// backstop that makes a repeat import of the same watch-recorded run
	// impossible rather than merely unlikely. Stored as given, not required
	// to be non-nil for SourceHealthKit — this module stores what it is told
	// (see ElevationGainM's note above), and enforcing that pairing is the
	// client's job, not a storage-level invariant.
	HealthKitUUID *string `json:"healthkit_uuid"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// Validate reports whether this is a detail record the system can store.
func (d SessionDetail) Validate() error {
	if !d.Source.Valid() {
		return ErrInvalidInput
	}
	if d.DistanceM != nil && *d.DistanceM < 0 {
		return ErrInvalidInput
	}
	if d.DurationSeconds != nil && *d.DurationSeconds < 0 {
		return ErrInvalidInput
	}
	if d.ElevationGainM != nil && *d.ElevationGainM < 0 {
		return ErrInvalidInput
	}
	if d.AvgPaceSecPerKm != nil && *d.AvgPaceSecPerKm < 0 {
		return ErrInvalidInput
	}
	if d.HealthKitUUID != nil && (*d.HealthKitUUID == "" || len(*d.HealthKitUUID) > maxHealthKitUUIDLength) {
		return ErrInvalidInput
	}
	if len(d.RoutePoints) > MaxRoutePoints {
		return ErrInvalidInput
	}
	for _, p := range d.RoutePoints {
		if !p.valid() {
			return ErrInvalidInput
		}
	}
	if len(d.Splits) > MaxSplits {
		return ErrInvalidInput
	}
	for _, s := range d.Splits {
		if !s.valid() {
			return ErrInvalidInput
		}
	}
	return nil
}

// Repository is the persistence port for the running half of a session.
//
// PutDetail/GetDetail, matching bjj.SessionRepository's shape exactly:
// PutDetail upserts and replaces the route/splits wholesale, so a retry
// after a partial failure (the mobile outbox's ordinary case) converges
// instead of duplicating rather than needing a merge.
type Repository interface {
	PutDetail(ctx context.Context, userID string, d SessionDetail) (SessionDetail, error)
	GetDetail(ctx context.Context, userID, sessionID string) (SessionDetail, error)
}
